### ISSUE TITLE
scepserver: comptibility with another SCEP clients

### ISSUE THREAD

**Issue by msva** [micromdm/scep#80]
Hi there!
It seems like either `scepserver` doesn't support certificate issuance by the requests from another SCEP client implementations, or I'm missing something important.

When I trying to issue a certifiate with `scepclient` implementation from this repo, it requests certificates just fine.

But when I trying to use some alternative software (where it is impossible to use client from this repo) I facing some problems.

Here is a part of the *server*'s log (with enabled `-debug`) when such a client (say, certmonger) tries to issue a cert:

```
level=info ts=2018-02-21T14:58:09.80919166Z caller=service_logging.go:22 component=scep_service method=GetCACaps err=null took=1.061µs
level=info ts=2018-02-21T14:58:09.809846342Z caller=endpoint.go:186 op=GetCACaps error=null took=656.663µs
level=info ts=2018-02-21T14:58:09.809911737Z caller=logutil.go:70 component=http method=GET status=200 proto=HTTP/1.1 host=[...] user_agent= path="/scep?operation=GetCACaps&message=0"
level=info ts=2018-02-21T14:58:09.941931207Z caller=service_logging.go:46 component=scep_service method=PKIOperation err="asn1: structure error: tags don't match (16 vs {class:1 tag:13 length:73 isCompound:false}) {optional:false explicit:false application:false defaultValue:<nil> tag:<nil> stringType:0 timeType:0 set:false omitEmpty:false} contentInfo @2" took=67.839µs
level=info ts=2018-02-21T14:58:09.941995663Z caller=endpoint.go:186 op=PKIOperation error=null took=133.181µs
level=info ts=2018-02-21T14:58:09.942030467Z caller=logutil.go:70 component=http method=GET status=500 proto=HTTP/1.1 host=[...] user_agent= path="/scep?operation=PKIOperation&message=MIIK%2BAYJKoZIhvcNAQcCoIIK6TCCCuUCAQExDzANBglghkgBZQMEAgEFADCCBZQGCSqGSIb3DQEHAaCCBYUEggWBMIIFfQYJKoZIhvcNAQcDoIIFbjCCBWoCAQAxggJSMIICTgIBADA2MDExCzAJBgNVBAYTAlVTMRAwDgYDVQQKEwdzY2VwLWNhMRAwDgYDVQQLEwdTQ0VQIENBAgEBMA0GCSqGSIb3DQEBAQUABIICAGewWOKoR2yg5%2BFSEgXjsUkjhf32zDD8xgwSIYuvzREHYQPvMyS7QNIMkXdtW546B0Qh8DF5Xo1PeE%2Fhoss3HC%2FaUK8NsDAi61EBJCrWneYkNfL9Z02YKMPXAOJWP4xn9J%2FqWIU6613Uce9C%2BFH9w1w6I5Ym1opMLdlR%2FpNfcoWY4B1YxVjWrVqOi1h7qdNp7WPozz8rD%2BSReXZn%2FF8uX4s8A%2FygFSOLZuM5sDgld%2BZyZoDFgqo2waryxcMeT4xcqGs0n9OlKATx4mOt1uiEuejnMIk3XeK0bKEsO2Ox5pTv%2FoIrL4jD9u7t9Ta%2BUjkVNlMV2hIuNdVTb4G0WlYVcrwbOqoJLFvu02KjULYrXDNjbrgAgJJfEYvMJ4AFuH9WtorhI%2FWCnm6Z1CV1j8HvyPVFwPJ5vhZcmRUz5cXrbEK9Q7PearJUT0hvG7H9baiX6enZ9a3rQ6VPwMZ1maXWFht%2BhXpdVsclJhKsb6D3LdJyolyq3MwsL4JH8D1fNLYGXE91c1zCF0Y9IjwA1caKUN%2BuNLFlY%2FdsMjBmj96l5o%2BR%2F%2Btf%2FGJhek6WcctunZnFVpsJ7d4WOrga%2BIyBgDKEfKh1LpQCXt5FVsI1OkbJkMABzccNKPIjCXHGo6yPh7c0Y5AbEhcLaF2tyZg6Cj7EeiVYBClCaqHPq%2BSV5Hmvf%2FOgMIIDDQYJKoZIhvcNAQcBMBQGCCqGSIb3DQMHBAiWYzY86ViajICCAujgc09L%2FKA6l8whqIrpljFrOYnrnJw%2B1ar59PkMv7m83fRKoB7axGIqy7YHpKWqZSPjQ5rzDwW1XzwDQR%2Fu%2B%2B%2Bpzj4ZPBTaMtXA1qEuhJV3%2BqCzy5%2BvroQ%2FGVjy9XjGWWiJdBbs7NZlvdL%2B8CtzRI7p0X8I%2By2Pn0EJ0jaGZeI4YP%2B7ihijXSYDBw9LjC%2FFtzyyvxNcomtHpB2VdReqMfMWor8j2lcVeYZLRYCocOfQHaow1n7tm7cNBKqIxGrEA69tXVd7E%2BbFFqRYXX%2B9rvsKRRTdwT9nJ%2FXwbVi7Us0ZkvCQOH8ru%2F31cCsVRMAswstYv9hJ9drGyNvMUVS%2BSV5gL9Sr5ARYGZjZv%2F5LYMZhyAPSdcQfOupnsuVjYkRQEq7PiOqHQnb3peTRbYxdh0tggwwyjbdA551AzX1h4x6z3DhVUL0DcN1Cv%2F1elxlCxtgRtImYTc7kKvuCZAceM7g9ns6TT1PyJXVeHoyfapOYFkKorUWgYBSn%2BDYeKQbE%2Fsk98pSaOT9H9a2K70W6UjVhq%2FOi7zRm6YANEOCFdfkIqXjvhE0D9QWNhN3gePu6jjFcUkbkHmcv0uZHth6WosfyWHYS1l4UP8MGzmiBUoHMwolIZiryeFe2h5EHNoWdKEcc9o4WVzts%2B4XFu%2BAU8UGnMfm4WXaBlqvqsDMydKgNu9I56NebjHB%2B1oVFuAVW3%2BKeFVx34Tq%2F5Myn5B7lkolZgWFafYb2texmpzFLQGAo7JA2q5X02jmShcYtMNUjYKyOFGMetjZSM5nPGuLxedLnsHmuB5Y6H%2FSYxkwmpNzVcEFFm8aGav6QVnitPS2pPrvfYsMgQVUneD7yMf9dgx36pgYWRCUSQr7YuvU8%2B0F6YQ2H1S%2Bz7ei3HCG%2Bg9b7cdidwToM83e5KQz26qNC%2FFb3rXs9DxN0RM1t0lIyeQzbb06ZfuL7l023VxbBvLILq122xHKqKxbhzvcYRRPCGedAU3r%2B6rh%2BpjagggLMMIICyDCCAbCgAwIBAAIgA6AyHuBaREMkvg%2B8vjIFtwYpeHr9J8VPnwkFB%2BmhsOUwDQYJKoZIhvcNAQELBQAwFjEUMBIGA1UEAxMLZnJvbG92LXRlc3QwIhgPMjAxODAyMjEwMDAwMDBaGA8yMTE4MDIyMTAwMDAwMFowFjEUMBIGA1UEAxMLZnJvbG92LXRlc3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDbf7tK67a0JwnQD1B5mCUljteG%2FCvFC6Wfpe0YtZt0A%2FLL%2B3pXMJjjM50Xy7Xn7k38kVaiEP0m%2FHyc0f9EzOTCnFQ5A6ocZc%2FifiN2WQZ4HbwMSo2XK%2FqXItXzIGy6AdUtfUUpfZxlw0IKOX%2BOWiw80W9ebq8KaINxorY4Ja78hHC6VxctlLvgWZR0poBDMAt61xpROxyMDAbhdwgr%2BWNxWSfRbqANn0%2ByXSKQPXELhgDPC3Kph%2FH%2FHelK%2FB4J7d19VxeTMxE7fmFukfGBbjDeHdUng8vzIFjU7Np0Vdf2MU0RakceCVNKVIri1o%2BAY%2Frr5lEUsIJ%2FxeaIoYFZLQKFAgMBAAEwDQYJKoZIhvcNAQELBQADggEBAIPwYIZh8XxT0cGwqSm5uGC%2BPGzudpD7Kc7kcCrqvp8k0TFgDhG%2BZT8GDPZu6sBP95AIM9cq3nvXrdZWQKoHf9gTBMf1p2g%2BWy3yqK8TgKs0Dw4eDtr2uOKCOtGH7XlumSqL0gVzyxppCDIJJjNFkuNwens1Z9%2FChhqPf408elirfyIsMv4I%2BWNDedoY%2BJmB47bkUqfNTrfQtT7BZ0WzNYlAblPoFdDL85GLkTAOxheqfB2sxyj1NeYPiGyZLZYCaEajrHcDKHau5ioAROtu2Lk%2FdudaSUSCTqKO80EWaU7sTQzrhgcocvwbLdgCojbURevIXaskr1lxIwSyKZAkPUExggJlMIICYQIBATA6MBYxFDASBgNVBAMTC2Zyb2xvdi10ZXN0AiADoDIe4FpEQyS%2BD7y%2BMgW3Bil4ev0nxU%2BfCQUH6aGw5TANBglghkgBZQMEAgEFAKCB%2FTASBgpghkgBhvhFAQkCMQQTAjE5MBgGCSqGSIb3DQEJAzELBgkqhkiG9w0BBwEwHAYJKoZIhvcNAQkFMQ8XDTE4MDIyMTE0NTgwOVowIAYKYIZIAYb4RQEJBTESBBDy5V7qp5Push%2BuXPcHs6nyMC8GCSqGSIb3DQEJBDEiBCCKl6y3xcoXICmrtpxHKqaz73og92Q6l63U91GXetNu1DBcBgpghkgBhvhFAQkHMU4TTDE2Mzk5Nzk5OTU4NTcxNjcwODAxMTIxNTAyMDExMjI0MjEyNDU3NjE0NTI5MjU1MzA4MzU3NDE2NDA3MDc5NTAwODY2MTY4ODc1MjUwDQYJKoZIhvcNAQEBBQAEggEAgiIzhNmI7YY3Dse%2FdACZMNDHmqUDmeDbLoZrVjwDWBj3qoRcfaPkgNqzGOehwQDT19pnTTVq7WxvbrbZG4PaoL%2F5gI3v72xql9dUkIqvLyEKlJLN7yA9x5NeaxI1xy6CsIrxeB%2Bm0uKvCavEHFjTTo%2FvrAgJZ0RIf1dMkLRw3IOToZBRXidbS8kEhT9AJSKWyQCwmypAWdtQcY0brFB4ua7yuPXHpR9K%2BADwocZPgfOjSSjp8tqcrUZCvUtD3IL7ETv5Px6%2F0e%2FeVgF5zpSLSJQ%2FPK6nrJjqIjz1F%2B1XQLEWB78L8AoqATyNP928mo7oELndqjnAdBEEjjnDiXkrEA%3D%3D"
```

As you can see, server somewhy throws a 500 error instead of successful certificate issuance.

On the client side (certmonger) it looks like:
```
Request ID '20180221145809':
status: CA_UNREACHABLE
ca-error: Server reply was of unexpected MIME type "text/plain".
stuck: no
key pair storage: type=FILE,location='/root/pki-test3/key.pem'
certificate: type=FILE,location='/root/pki-test3/cert.pem'
signing request thumbprint (MD5): F21ADF60 A76D9342 BE0088F9 2AE5064B
signing request thumbprint (SHA1): 89A1B707 B63A8281 7283BF46 9A4DD55D 7EE47064
CA: test1
issuer: 
subject: 
expires: unknown
pre-save command: 
post-save command: 
track: yes
auto-renew: yes
```

And there is also glitches with Mikrotik's client:
```
[admin@MikroTik] /certificate> print detail 
Flags: K - private-key, D - dsa, L - crl, C - smart-card-key, A - authority, I - issued, R - revoked, E - expired, 
T - trusted 
0 K T name="testcert1" issuer=CN=tik-test2 common-name="tik-test2" key-size=2048 days-valid=365 trusted=yes 
key-usage=key-cert-sign scep-url="http://12.34.56.78:8080/scep" serial-number="07F48C47E6CD4C55" 
fingerprint="f8add6a69dce5c09268a31b486207d57916f56bc387d88c67d82697a2467252f" 
ca-fingerprint="11d77808b90fb4732a4a45892680d6789a26d8deb47e35d97d82ed70fb3a25f3" 
invalid-before=feb/21/2018 17:55:21 invalid-after=feb/21/2019 17:55:21 challenge-password="secret" 
status="requesting-certificate-failed"
```

While server logs shows something like that on Mikrotik request:
```
level=info ts=2018-02-21T15:06:54.474679496Z caller=service_logging.go:22 component=scep_service method=GetCACaps err=null took=1.928µs
level=info ts=2018-02-21T15:06:54.474844924Z caller=endpoint.go:186 op=GetCACaps error=null took=170.369µs
level=info ts=2018-02-21T15:06:54.474908144Z caller=logutil.go:70 component=http method=GET status=200 proto=HTTP/1.1 host=[...] user_agent="Mikrotik/6.x SCEP" path="/scep?operation=GetCACaps&message=DummyCAIdentity"
level=info ts=2018-02-21T15:06:54.586631769Z caller=service_logging.go:34 component=scep_service method=GetCACert err=null took=2.044µs
level=info ts=2018-02-21T15:06:54.58676655Z caller=endpoint.go:186 op=GetCACert error=null took=139.792µs
level=info ts=2018-02-21T15:06:54.58681158Z caller=logutil.go:70 component=http method=GET status=200 proto=HTTP/1.1 host=[...] user_agent="Mikrotik/6.x SCEP" path="/scep?operation=GetCACert&message=DummyCAIdentity"
level=debug ts=2018-02-21T15:07:03.238885117Z caller=scep.go:244 msg="parsed scep pkiMessage" scep_message_type="PKCSReq (19)" transaction_id=317be4cdbb86b220746909e6224bde79bfa5e8a19bf340cc7ebc3ee217a62b61
level=debug ts=2018-02-21T15:07:03.253940734Z caller=scep.go:328 msg="decrypt pkiEnvelope" encryption_algorithm=0 has_challenge=true
level=info ts=2018-02-21T15:07:03.280811676Z caller=service_logging.go:46 component=scep_service method=PKIOperation err=null took=42.113125ms
level=info ts=2018-02-21T15:07:03.280835701Z caller=endpoint.go:186 op=PKIOperation error=null took=42.138907ms
level=info ts=2018-02-21T15:07:03.280881016Z caller=logutil.go:70 component=http method=POST status=200 proto=HTTP/1.1 host=[...] user_agent="Mikrotik/6.x SCEP" path="/scep?operation=PKIOperation&message=DummyCAIdentity"
```

**Comment by groob** [comment]
thanks for the detailed bug report. The issue is that every implementation of SCEP is slightly different, so depending on the client the formats being sent to the server are not going to be exactly the same. The `scepserver` here is compatible with Apple's built in SCEP client so that macOS and iOS devices can requests certs. 

I'm happy to extend support for other clients or accept pull requests which add support. 

For the certmonger example, it's failing because the PKIOperation envelope is not encoded in a way that scepserver can parse. Could you capture the full request sent with PKIOperation?

**Comment by groob** [comment]
With microkit, It looks like the request succeeded as far as the server is concerned? The logs from the client are not very useful, there's no reason given why it failed.

**Comment by groob** [comment]
For additional context, `micromdm/scep` implements [this](https://tools.ietf.org/html/draft-gutmann-scep-08) version of the SCEP RFC, but there are about 30 other versions of this protocol implemented over 30 years, so ¯\_(ツ)_/¯

**Comment by yoshiotu** [comment]
I looked into this a little for certmonger v. 0.78.6 client on ubuntu 16.04.

For GET operation=PKIOperation, certmonger client is base64-encoding the message query option. It seems to decode successfully using Go's base64.StdEncoding

certmonger client then fails because the reply has no sender nonce, which the current ietf draft says is superfluous. I tried making scepserver stick on a nonce, and then certmonger client worked.

Maybe the client is following this old specification:  https://tools.ietf.org/html/draft-nourse-scep-18
